libp11: new recipe

dacianf opened the pull request.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,69 @@
+COPYRIGHT="1991, 1999 Free Software Foundation"
+LICENSE="GNU LGPL v2"
+REVISION="1"
+SOURCE_URI="$HOMEPAGE/releases/download/libp11-0.4.3/libp11-0.4.3.tar.gz"
```
You can change SOURCE_URI to:
```
SOURCE_URI="$HOMEPAGE/releases/download/libp11-$portVersion/libp11-$portVersion.tar.gz"
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,69 @@
+	devel:pkcs11$secondaryArchSuffix = $portVersion
+	"
+REQUIRES_devel="
+	libp11$secondaryArchSuffix == $portVersion
```
We need the "base" keyword at the end:
```
	libp11$secondaryArchSuffix == $portVersion base
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,69 @@
+INSTALL()
+{
+	make install
+	prepareInstalledDevelLibs \
```
We like to remove the libtool files. We do that after the "make install" with:
```
	rm $libDir/libp11.la $libDir/pkcs11.la
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,69 @@
+SOURCE_DIR="libp11-$portVersion"
+
+ARCHITECTURES="x86_gcc2 ?x86 ?x86_64"
+SECONDARY_ARCHITECTURES="x86"
```
We will need to change SECONDARY_ARCHITECTURES (and maybe ARCHITECTURES too) but this will depend on which architectures the recipe builds or fails to build. So I'll update this comment later.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,69 @@
+SUMMARY="A library that implement a layer on top of PKCS#11 API"
```
"that implement"  -> "that implement**s**" or maybe "implementing"

BTW, I think we can improve DESCRIPTION too, but don't worry about this.
If anyone has any suggestion, please do not hesitate!

humdingerb replied:
As DESCRIPTION I'd suggest to take it directly from the homepage. 1st paragraph of "PKCS#11" and then the "Overview":

> The PKCS#11 API is an abstract API to access operations on cryptographic objects such as private keys, without requiring access to the objects themselves. That is, it provides a logical separation of the keys from the operations.
> The PKCS #11 API is mainly used to access objects in smart cards and Hardware or Software Security Modules (HSMs). That is because in these modules the cryptographic keys are isolated in hardware or software and are not made available to the applications using them.
> 
> There are two libraries:
> 
> - libp11 provides a higher-level (compared to the PKCS#11 library) interface to access PKCS#11 objects. It is designed to integrate with applications that use OpenSSL.
> - pkcs11 engine plugin for the OpenSSL library allows accessing PKCS#11 modules in a semi-transparent way.
> 
> The wiki page for this project is at https://github.com/OpenSC/libp11/wiki and includes a bug tracker and source browser.

dacianf commented:
@fbrosson what do i have to change here?

fbrosson commented:
@dacianf: I don't know what needs to be changed.
Did you manage to build libp11 or libp11_x86?
This is what I get, when INSTALL is called:
```
Collecting files to be packaged ...
Making install in src
make[1]: Entering directory '/sources/libp11-0.4.3/src'
make[2]: Entering directory '/sources/libp11-0.4.3/src'
 /bin/mkdir -p '/packages/openssl-1.0.2j-1/.self/develop/lib/engines'
/bin/mkdir: cannot create directory '/packages/openssl-1.0.2j-1/.self/develop/lib/engines': Read-only file system
make[2]: *** [Makefile:470: install-enginesexecLTLIBRARIES] Error 1
make[2]: Leaving directory '/sources/libp11-0.4.3/src'
make[1]: *** [Makefile:860: install-am] Error 2
```
Shall I add the "help-wanted" label on this PR?

dacianf commented:
@fbrosson It builds successfully for me on both arch :)

fbrosson commented:
That's interesting. Maybe the problem comes from my build of openssl or something else in my system.
Which version of openssl do you have?
Would you mind posting somewhere (maybe to a pastebin) the logs of your build of libp11 (and maybe libp11_x86 too)?
Do these commands:
```
haikuporter --get-dependencies --test libp11
haikuporter --get-dependencies --test libp11_x86
```
provide any interesting output, I mean, does libp11 come with a real "check" target in the Makefile?

Thanks!

(I'll restore the original title of this PR. I though something was wrong, but if that's only on my system then I guess the PR will be merged by someone else ;-)

dacianf commented:
@fbrosson  output of first command http://pastebin.com/waNHHQyL and second one: http://pastebin.com/YydrkGY8

fbrosson commented:
Thanks @dacianf for the test logs.
It looks like we need SoftHSM to test libp11. We had a task for creating a recipe for it, put it was not published because softhsm depends on botan. We also had a task for creating a recipe for botan, but it was not claimed...

Hmm, if you enjoyed writing recipes during the GCI and wish to continue outside of the GCI, here are some pointers and information to create the recipes for botan and softhsm, which will be useful for testing libp11 ;-)

**`haikuports/dev-libs/botan/botan-2.0.1.recipe`**
* SUMMARY="A crypto library written in C++"
* HOMEPAGE="https://botan.randombit.net/"
* SOURCE_URI="https://github.com/randombit/botan/archive/2.0.1.tar.gz"
* SOURCE_FILENAME="botan-$portVersion.tar.gz"

**`haikuports/dev-libs/softhsm//softhsm-2.2.0.recipe`**
* SUMMARY="A software PKCS#11 implementation from the OpenDNSSEC project"
* HOMEPAGE="https://www.opendnssec.org/softhsm/"
* SOURCE_URI="https://github.com/opendnssec/SoftHSMv2/archive/2.2.0.tar.gz"
* SOURCE_FILENAME="softhsm-$portVersion.tar.gz"

BTW, the letters "HSM" (in SoftHSM) stand for Hardware Security Modules.

But please do not bother with this if you don't really want to or if you don't have enough time.
There is no urgency and we will also have a new web page with a list of recipes taken from all the ones that were not claimed during the GCI. So botan and softhsm will have their recipes sooner or later ;-)

dacianf commented:
@fbrosson I will try to do these recipes.

dacianf commented:
@fbrosson I get this error when i try to build botan_x86 http://pastebin.com/KpDTdhSE

fbrosson commented:
Wow, that's great! You already wrote a recipe for botan!
It looks OK, except for the call to configure.py.
I think we need to replace it by
```
	runConfigure ./configure.py
```
Hmm, it does not work either. I'm trying some variants right now.

fbrosson commented:
@dacianf, I have good news: it looks like your recipe for botan is building fine (with a few minor changes, as shown included in my previous message.)
I don't think this is necessary, but I've added:
```
	devel:libbz2$secondaryArchSuffix
	devel:libcrypto$secondaryArchSuffix
	devel:liblzma$secondaryArchSuffix
	devel:libsqlite3$secondaryArchSuffix
	devel:libssl$secondaryArchSuffix
	devel:libz$secondaryArchSuffix
```
to BUILD_REQUIRES.

And I've also added a few options to the configure script:
```
	runConfigure --omit-buildspec ./configure.py \
		--with-bzip2 --with-lzma --with-zlib \
		--with-openssl --with-sqlite3
```
After the build finishes, I'll see what lib:lib\* lines need to be added to REQUIRES.

dacianf commented:
It still fails for me...

2017-01-17 23:40 GMT+02:00 fbrosson <notifications@github.com>:

> @dacianf <https://github.com/dacianf>, I have good news: it looks like
> your recipe for botan is building fine (with a few minor changes, as shown
> included in my previous message.)
> I don't think this is necessary, but I've added:
>
> 	devel:libbz2$secondaryArchSuffix
> 	devel:libcrypto$secondaryArchSuffix
> 	devel:liblzma$secondaryArchSuffix
> 	devel:libsqlite3$secondaryArchSuffix
> 	devel:libssl$secondaryArchSuffix
> 	devel:libz$secondaryArchSuffix
>
> to BUILD_REQUIRES.
>
> And I've also added a few options to the configure script:
>
> 	runConfigure --omit-buildspec ./configure.py \
> 		--with-bzip2 --with-lzma --with-zlib \
> 		--with-openssl --with-sqlite3
>
> After the build finishes, I'll see what lib:lib* lines need to be added to
> REQUIRES.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/haikuports/haikuports/pull/1046#issuecomment-273308607>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASodEgAmrLmHgO9uTuTJWRHjn7nOhdF2ks5rTTU5gaJpZM4Lhswt>
> .
>

dacianf commented:
@fbrosson it says makefile 1691 recipe target install failed

fbrosson commented:
(Sorry @dacianf for this late reply. I was AFK for a long time.)

I think I got the same error as you for botan.
BTW, I tried to investigate what is causing this error:
```
Read-only file system: '/packages/botan_x86-2.0.1-1/.self'
```
I have a similar error when building libp11.
Hmm, what was the hrev of your system when you successfully built libp11?
Did ypu upgrade your system (with pkgman update) since your last successful build of libp11?
If yes, which hrev are you using now? And does libp11 still build fine?

I need to check if tpackagefs has changed since a few weeks and see if I can build libp11 or botan with an old hrev.

BTW, in case you try to rebuild libp11, could you post the logs to a pastebin?
Thanks!

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,70 @@
+	devel:pkcs11$secondaryArchSuffix = $portVersion
+	"
+REQUIRES_devel="
+	libp11$secondaryArchSuffix == $portVersion base
```
Oops, I just noticed we forgot to insert
```
	haiku${secondaryArchSuffix}_devel
```
in REQUIRES_devel, before
```
	libp11$secondaryArchSuffix == $portVersion base
```
(BTW, I hoped this would make my build of libp11 succeed, but it did not. The problem is probably somewhere else in my system.)

dacianf commented:
@fbrosson here is log of libp11 building http://pastebin.com/zrwhdNyR

fbrosson commented:
Thanks @dacianf for the update and for your build log.
I can see that you are on Haiku(x86_gcc2) hrev50883, while I'm 7 hrevs behind, so I'll update, just in case.
I can also see that you have openssl-1.0.2g, while I have openssl-1.0.2j.
But the recipes of openssl 1.0.2g and 1.0.2j are almost identical (1.0.2j has cmd:cmp in BUILD_PREREQUIRES), so it is unlikely that my problem could come from openssl 1.0.2j.

Regarding the addition of ${secondaryArchSuffix}_devel in REQUIRES_devel (which you made on my request), I noticed (afterwards) that most packages don't have it, so I guess it was not really needed, but it does not harm so there is no need to remove it.

Has anyone else tried to build libp11? If it works fine, please feel free to merge this recipe!

dacianf commented:
@fbrosson What about botan, it doesn't build for me....

fbrosson commented:
botan is also failing to build on my system... and I can't find what's wrong ;-(
Let's hope we will get help from some HaikuPorts experts (for libp11 and maybe for botan too ;-)
BTW, I noticed that your remote libp11_recipe branch is 65 commits behind master.
This is not required, but in case you would like to update your local libp11_recipe branch and rebase it against master, I can try to guide you. This can be handy if you have a branch associated to a PR that has not been merged and you wish to build newly commited recipes without having to switch to/from your different branches.
Once you are comfortable with updating your local branches (against the project's master), you will probably want to create a PR for botan.  That would be great, BTW, since it will be easier to suggest changes to your recipe. (I already have a copy of your recipe for botan, taken from your last pastebin and updated with my suggestions, so there is no need to create a PR right now, unless someone else would like to review it ;-)

scottmc commented:
I would mark it as broken on the platforms that it is failing to build and/or run on, and then check it in.  Then open a new issue to report the errors you are seeing so if/when someone else picks up work on it they have some point to start.

fbrosson commented:
@scottmc, if you had libp11 in mind, well, it **is** building fine on @dacianf's system. It's just on my system that INSTALL is failling. So I guess we should not mark them as broken.
I cannot merge this PR myself because I haven't been able to build it.
But I hope someone else will test and merge this PR.

But maybe your message was about botan.
@dacianf hasn't created a PR for it yet, His recipe for it, with my suggestions, is still not building (for both of us.)
From your message I understand that it would still be nice to have a PR for it, even if it does not build, as long as we mark each architecture as broken.

dacianf commented:
@fbrosson can you help me with updating my branch?

fbrosson commented:
Sure!
First I need to know what are the nicknames of your remote repos.
You can get that info with:
```
git remote -v
```
In case you have a somilar config as me, you would see:
```
origin	https://github.com/dacianf/haikuports.git (fetch)
origin	https://github.com/dacianf/haikuports.git (push)
upstream	https://github.com/haikuports/haikuports.git (fetch)
upstream	https://github.com/haikuports/haikuports.git (push)
```
in which case you can try this:
```
git fetch upstream master
git merge --no-edit upstream/master
git rebase upstream/master
```
But before you try this on your libp11_recipe branch, I would recommend that you try it on a newly created copy of your libp11_recipe branch, that we can call libp11_updated:
```
git checkout -b libp11_updated libp11_recipe
git fetch upstream master
git merge --no-edit upstream/master
git rebase upstream/master
```
If everything is OK, you can either:

1) switch back to the libp11_recipe branch
```
git checkout libp11_recipe
```
and run the 3 commands again on your libp11_recipe branch:
```
git fetch upstream master
git merge --no-edit upstream/master
git rebase upstream/master
```
and then push it:
```
git push -f
```

2) **or**, if you prefer, rename libp11_recipe to libp11_recipe_old and libp11_updated to libp11_recipe:
```
git branch -M libp11_recipe libp11_recipe_old
git branch -M libp11_updated to libp11_recipe
git push -f --set-upstream origin libp11_recipe
```

Reminder: you don't need to push your branches everytime you rebase them. I only do that if I have changes that I need to push to the PR itself.
But this is OK to push it after a rebase. GitHub will show that you have updated your commit even if you have just rebased it, but that's OK.

If "git remote -v" yelds other nicknames, then you'll need to tell me what their names are, so that I can send you the correct commands.

fbrosson commented:
@dacianf, I found what was wrong in the current recipe for libp11:
"make install" tries to copy pkgs11.so into the $libDir/engines/ sub-directory of the... openssl package.
But that directory is read-only, so we need to use the $libDir/engines/ of libp11.
Passing --with-enginesdir=$libDir/engines to configure allows to do that.
I'm doing some polishing and then I'll add a review.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+SUMMARY="A library that implements a layer on top of PKCS11 API"
```
Maybe
```
SUMMARY="A PKCS#11 wrapper library"
```
would be better, to avoid the repetition with the suggestion for DESCRIPTION.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+SUMMARY="A library that implements a layer on top of PKCS11 API"
+DESCRIPTION="The official header files for PKCS11 Standard were released \
+under a non-free license. libp11 does not include these header files, but \
+instead uses a rewritten pkcs11.h header file from the Scute project."
```
I think
```
DESCRIPTION="libp11 is a library implementing a thin layer on top of PKCS#11 \
API to make using PKCS#11 implementations easier."
```
would be better than
```
DESCRIPTION="The official header files for PKCS11 Standard were released \
under a non-free license. libp11 does not include these header files, but \
instead uses a rewritten pkcs11.h header file from the Scute project."
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+under a non-free license. libp11 does not include these header files, but \
+instead uses a rewritten pkcs11.h header file from the Scute project."
+HOMEPAGE="https://github.com/opensc/libp11"
+COPYRIGHT="1991, 1999 Free Software Foundation"
```
Should be
```
COPYRIGHT="2016 Michał Trojnara
	2015-2016 Nikos Mavrogiannopoulos
	2007-2009 Andreas Jellinghaus
	2011 Martin Paljak"
```
Most active authors appear first.
Source: [NEWS](https://github.com/OpenSC/libp11/blob/master/NEWS) file from the latest archive.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+instead uses a rewritten pkcs11.h header file from the Scute project."
+HOMEPAGE="https://github.com/opensc/libp11"
+COPYRIGHT="1991, 1999 Free Software Foundation"
+LICENSE="GNU LGPL v2"
```
Should be
```
LICENSE="GNU LGPL v2.1"
```
( "2" -> "2.1")

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+SOURCE_DIR="libp11-$portVersion"
+
+ARCHITECTURES="x86_gcc2 ?x86 ?x86_64"
+SECONDARY_ARCHITECTURES="x86"
```
Please replace
```
ARCHITECTURES="x86_gcc2 ?x86 ?x86_64"
SECONDARY_ARCHITECTURES="x86"
```
by
```
ARCHITECTURES="x86_gcc2 x86 x86_64"
SECONDARY_ARCHITECTURES="x86_gcc2 x86"
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+	libp11$secondaryArchSuffix = $portVersion
+	lib:libp11$secondaryArchSuffix = $portVersion
+	lib:libpkcs11$secondaryArchSuffix = $portVersion
+	lib:pkcs11$secondaryArchSuffix = $portVersion
```
Please replace
```
	lib:libp11$secondaryArchSuffix = $portVersion
	lib:libpkcs11$secondaryArchSuffix = $portVersion
	lib:pkcs11$secondaryArchSuffix = $portVersion
```
by
```
	lib:libp11$secondaryArchSuffix = 2.4.3 compat >= 2
	lib:libpkcs11$secondaryArchSuffix
	lib:pkcs11$secondaryArchSuffix
```
BTW, **`libpkcs11.so`** is just a symlink to **`pkcs11.so`**, so maybe we should drop
```
	lib:libpkcs11$secondaryArchSuffix
```
I tried with and without that line, but "hp -S" did not complain with any of these variants.
Hmm, maybe it does not complain about libs which are not in $libDir. (That lib is in $libDir/engines/.)

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+	libp11${secondaryArchSuffix}_devel = $portVersion
+	devel:libp11$secondaryArchSuffix = $portVersion
+	devel:libpkcs11$secondaryArchSuffix = $portVersion
+	devel:pkcs11$secondaryArchSuffix = $portVersion
```
Please replace
```
	devel:libp11$secondaryArchSuffix = $portVersion
	devel:libpkcs11$secondaryArchSuffix = $portVersion
	devel:pkcs11$secondaryArchSuffix = $portVersion
```
by
```
	devel:libp11$secondaryArchSuffix = 2.4.3 compat >= 2
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+	devel:libcrypto$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:awk
```
Please insert
```
	cmd:aclocal
	cmd:autoconf
	cmd:autoheader
	cmd:automake
```
before in BUILD_PREREQUIRES before cmd:awk.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+
+BUILD()
+{
+	runConfigure ./configure
```
Should be
```
	runConfigure ./configure --with-enginesdir=$libDir/engines
```
otherwise "make install" tries to copy pkcs11.so in the $libDir/engines/ directory of the openssl package.
But that directory is read-only, so we need to tell libp11 to install its pkcs11.so in its own $libDir/engines/.

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,71 @@
+	rm $libDir/libp11.la $libDir/pkcs11.la
+	prepareInstalledDevelLibs \
+		pkcs11 \
+		libp11
```
Please replace
```
	rm $libDir/libp11.la $libDir/pkcs11.la
	prepareInstalledDevelLibs \
		pkcs11 \
		libp11
```
by
```

	rm $libDir/libp11.la
	rm $libDir/engines/pkcs11.la

	prepareInstalledDevelLib libp11
	fixPkgconfig

```
(keeping a blank line above the first rm and another one below the fixPkgconfig, so that the recipe looks nicer ;-)

fbrosson commented:
@dacianf, @scottmc: Good news: I managed to build botan!
We just need to pass --destdir=/ to ./configure.py.
@dacianf, after you fix the recipe for libp11 (with the comments from my [previous review](#pullrequestreview-17600700)), would you like to create a PR for botan?
Don't worry if you don't have time right now. We can wait, or, if you prefer, I can add botan myself.
Thanks!
BTW; it might be a good idea to leave this PR open until botan and softhsm become available, so that we can add softhsm as a dependency for libp11 to have the tests in libp11 really done instead of being skipped.

dacianf commented:
@fbrosson I will make a pr soon for botan,I've also made a recipe for softhsm but it shows some errors.

dacianf commented:
@fbrosson i've made that pr https://github.com/haikuports/haikuports/pull/1094

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,77 @@
+COPYRIGHT="2016 Michal Trojnara
+		2015-2016 Nikos Mavrogiannopoulos
+		2007-2009 Andreas Jellinghaus
+		2011 Martin Paljak"
```
Please reduce the indentation of lines 6 to 8 by one tab.

BTW, the L in "Michał" is not the "usual" lowercase L, it is a variant with an sort of very small slash over it.
Since HaikuPorts accepts unicode chars in recipes, it would be nice to use the correct name.
You can copy the correct lowercase L from this comment, as well as from my [this message](#discussion_r97002736), or from the NEWS file
Sorry for being picky about this small detail. Thanks!

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,77 @@
+
+PROVIDES="
+	libp11$secondaryArchSuffix = 2.4.3 compat >= 2
+	lib:libp11$secondaryArchSuffix
```
Please replace
```
	libp11$secondaryArchSuffix = 2.4.3 compat >= 2
	lib:libp11$secondaryArchSuffix
```
by
```
	libp11$secondaryArchSuffix = $portVersion
	lib:libp11$secondaryArchSuffix = 2.4.3 compat >= 2
```
(libp11 is the name of the package, while lib:libp11 is the lib itself. The first needs a $portVersion, while the  second needs the "2.4.3".)
BTW, I'm pretty sure is what will fix the error you are getting:
```
No such file or directory: '/boot/home/haikuports/dev-libs/libp11/work-0.4.3'
```

fbrosson commented on dev-libs/libp11/libp11-0.4.3.recipe:
```diff
@@ -0,0 +1,77 @@
+SUMMARY="A PKCS#11 wrapper library"
+DESCRIPTION="libp11 is a library implementing a thin layer on top of PKCS#11 \
+API to make using PKCS11 implementations easier"
```
PKCS11 -> PKCS#11
easier -> easier.
(missing full-stop after "easier")

fbrosson commented:
@dacianf, I assumed you were busy, so I pushed the fixes and merged the recipe.

BTW, now that this PR has been merged, the branch it is based on is no longer needed.
You should therefore click "Delete branch" for this PR to get rid of an unneeded branch in your repo. Thanks!
GitHub help: [Deleting unused branches](https://help.github.com/articles/deleting-unused-branches/)
